Time-Aggregate TimeseriesDatasets

atrejut opened the pull request:
First implementation of making time-aggregates of TimeseriesDatasets. 
Plotting a dataset needed some small adjustments to deal with the additional information on the aggregation function. `as_df` works out of the box, also with `speaking_names=True`. 
Tests will follow, but maybe take a general look already @cloudyday?

@JaakobKind FYI

cloudyday commented on sailor/sap_iot/wrappers.py:
```diff
@@ -365,3 +375,45 @@ def filter(self, equipment_ids: Iterable[str] = None, start: Union[str, pd.Times
+
+    def aggregate(self,
+                  aggregation_interval: str,
+                  aggregation_functions: Union[Iterable[Union[str, Callable]], str, Callable] = 'mean')\
```
looking at this type hint, I want to cry :D

cloudyday reviewed (commented):
Just looked at the `aggregate` function so far. looks good. need to try the plot soon

cloudyday commented:
Ah and what we missed in the last PR... this should also be a cached property. maybe you can add that here
https://github.com/SAP/project-sailor/blob/main/sailor/assetcentral/indicators.py#L71

cloudyday commented on sailor/sap_iot/wrappers.py:
```diff
@@ -235,17 +237,25 @@ def plot(self, start=None, end=None, indicator_set=None, equipment_set=None):
             aggregation_interval = '1h'
         else:
             aggregation_interval = '12h'
-        groupers = [pd.Grouper(key=time_column, freq=aggregation_interval), *self.get_key_columns()]
```
I just noticed that line 231 to 239 should not be used if we have an AggregatedIndicator. We should be able to just get the aggregation interval from it.

atrejut replied:
I don't think it's necessarily the same aggregation interval actually.
The interval here is chosen such that the graph looks good, nothing else. If the `AggregateIndicator` has a 2-min interval, and you want to plot data for say a month here you still want the plotting function to aggregate to a longer interval. [30 days/month * 24 hours/day * 30 (2-min-aggregates/hour) = 21600 points, bit too much for the resolution of my screen :)]

cloudyday replied:
Ah sure. Thanks for clearing that up

cloudyday reviewed (commented):
I also noticed the x axis scale is not what I would expect when I have a short timeframe (< 30 days) and an aggregation interval of 1d.

You'll then get labels with 2 days step size, which you would not expect with an aggregation interval of 1d.


But I guess this is actually not to be fixed here...maybe, maybe not.

atrejut commented:
Hi @cloudyday 

so I've refactored the aggregation and x-axis break calculation in plotting, and added a new helper function for that.
Also added tests for that new helper function, as well as at least some tests for aggregation (maybe more could be done there, but I don't have good ideas just now. suggestions welcome).

In the end there are quite some changes to the plotting by now. I've not added new tests here though, as testing plotting seems tricky -- will need to think about that separately I think.

cloudyday reviewed (approved):
You found a nice way to improve on the x-axis scaling. 

+1
